### PR TITLE
Add Menu and Dropdown design guidelines

### LINKED ISSUE #2115: Create dropdown design guidelines
Want to have some guidelines for drop downs, to describe how they work, how they are different from select lists (because I know this has been a point of confusion for people), and add in some examples!

Talk about how drop downs just list out a list of actions that you can take. It is different from a select list in that you don't select an option and that's it - selecting a menu item in a dropdown takes you somewhere else...

+ add anything else you think makes sense!

### LINKED ISSUE #2118: Create design guidelines for the Menu component
- Explain the different types of menu options that there are
- Mention the relationship between menu items and drop downs
- Talk about the different menu styles, and "nested" menu items (ie: menu items that can unveil more menu options from that category)

Look at react examples for more ideas

### PR DESCRIPTION
**Pull request by mcarrano**
This PR adds new design guidelines documents for the Dropdown and Menu components. It also includes updates to the Select design guidelines, adding an Elements and variations section before usage to make this more similar to the Dropdown guidelines page.

Closes #2115 
Closes #2118

### CODE REVIEW

File: packages/v4/src/content/design-guidelines/components/dropdown/dropdown.md [hunk 1]
```diff
@@ -2,4 +2,39 @@
 id: Dropdown
 section: components
 ---
-Use a **dropdown** when you want to present a list of actions in a limited space.
+
+A **dropdown** component is used to present a menu of actions or links in a constrained space that will trigger a process or navigate to a new location. See Select component [link] if you want to make one or more selections from a list of items in a value list.
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/dropdown/dropdown.md:6]: Need to edit the "select component [link]" to:
"See [Select component](/components/select/) if you want..."

Comment (mmenestr) [reply packages/v4/src/content/design-guidelines/components/dropdown/dropdown.md:6]: Also change beginning to "A **Dropdown** is used to..." - so capitalize dropdown, and remove the word "component". This is just so it's more in line with the start of other guidelines!

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
+---
+id: Menu
+section: components
+---
+
+A menu component that can be used in a variety of contexts whenever the user needs to choose between multiple values, options, or actions. A menu can be opened in a [dropdown](/components/dropdown), a [select list](/components/select), or it can be revealed upon right clicking on a region within a page.
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/menu/menu.md:6]: Change this too... "A **Menu** can be used in a variety...."
It doesn't really tell you what a menu is though, and more so how you can use it. Should we maybe add a sentence like "A **Menu** is a list of items inside a contained that can be used..." or something along those lines

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+## Elements and variations
+### Action menu
+An actions menu is used for presenting a list of actions or links.
+
+<img src="./img/basic-menu.png" alt="actions dropdown" width="863"/>
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/menu/menu.md:12]: I feel like the pink headers in this image should be in sentence case? I'm guessing you title cased it because it's referring to the bullets, but I think it looks odd since they're just general "statements"

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+3. **Links:** Links may be used in place of or mixed with actions to navigate the user to a new page. If the link will open in a new window, the external link icon should be used to annotate the link and let the user know to expect this behavior.
+
+4. **Descriptions (optional):** Descriptive text may be added below each menu item. This should only be used when the menu item label by itself may not be clear to all users. Keep descriptive text to two lines or less.
+
+### Option select menus
+Use an option select menu when you want to persist the selected items. This is typically the case for select menus or settings menus that allow users to select between multiple options. See the [Select](/components/select) and [Options menu](/components/options-menu) components for more details. You may present a single or multiple groups of options within the same menu.
+
+<img src="./img/option-select-menu.png" alt="option select menus" width="863"/>
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/menu/menu.md:25]: Are the checkboxes in these images meant to be grey? Or should they be blue? I know they are grey in the symbol library but I feel that this may be a bug. Been meaning to bring this up to Marie!

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+2. **Icon (optional):** You may place an icon before each menu item to aid in recognition. It has been shown that familiar icons speed the recognition of text labels, but only use icons if they will be easily recognized and are distinguishable. Never use icons simply as a decoration.
+
+3. **Links:** Links may be used in place of or mixed with actions to navigate the user to a new page. If the link will open in a new window, the external link icon should be used to annotate the link and let the user know to expect this behavior.
+
+4. **Descriptions (optional):** Descriptive text may be added below each menu item. This should only be used when the menu item label by itself may not be clear to all users. Keep descriptive text to two lines or less.
+
+### Option select menus
+Use an option select menu when you want to persist the selected items. This is typically the case for select menus or settings menus that allow users to select between multiple options. See the [Select](/components/select) and [Options menu](/components/options-menu) components for more details. You may present a single or multiple groups of options within the same menu.
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/menu/menu.md:23]: I think it would help to link "groups of options" to that section in the doc --> [groups of options](#grouped-menus)

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+### Filtering menu items
+If the list of possible items is very long, a filter may be added to make it easier for the user to find what they are looking for.
+
+<img src="./img/menu-filtering.png" alt="grouped menu" width="348"/>
+
+When filtering is being used, a search input will be added to the top of the menu and the list of items will be filtered as the user types.
+
+### Favoriting and other actions
+Favoriting or any other optional action can be associated with any menu item. In this case, making an item a favorite will duplicate in at the top of the menu. This may be useful when you have a long list of possible actions and want to “pin” a subset of items to the top of the menu.
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/menu/menu.md:57]: little typo in here - should say "will duplicate **it**" not "duplicate **in**"

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+<img src="./img/favorities-menu.png" alt="menu with favorites" width="348"/>
+
+
+Any action that can be represented as an icon button can be placed in one or more menu items. In this case the favorites action is used to mark an item as a favorite.
+
+## Usage
+
+| **When you want to...** |                     |
+| ------------------- | ------------------- |
+| Expose a list of actions in a dropdown menu | Use an actions menu with actions and/or links. Add icons in front of menu items only if they will help to make items more recognizable. Include descriptive text for menu items that may require further definition. |
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/menu/menu.md:68]: would take out the ellipses at the end of each use case, the first row doesn't have them, and some are alone on the next line which looks a little weird

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+| Create logical groupings of menu items... | Use separators to group items. Include group headings if you don’t feel like the meaning of the groups will be obvious. |
+| Expose a list of nested actions... | Use a fly-out menu. Any optional attributes of basic menus, including icons, groups, and descriptions can also be applied to flyout menus.|
+| Select from a list of options in a dropdown menu ... | Use an options select menu for selecting between values rather than actions. Icons, descriptions, and groupings can be used with option select menus to improve recognition or create logical option groups. |
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/menu/menu.md:71]: This last case is a little confusing, are you talking about select lists, or drop downs? It seems like you are talking about select lists here so I would probably remove any mention of "drop downs" to prevent confusion!

Would rephrase to "select from a list of options in a select menu..." "Use an options select menu etc"

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+## Content guidelines
+It creating menu item labels, keep in mind the following guidelines:
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/menu/menu.md:74]: Typo - assuming you meant "in creating" not "it creating"

File: packages/v4/src/content/design-guidelines/components/select/select.md [hunk 10]
```diff
@@ -9,7 +9,29 @@ related: [
...
+### Basic select
+A select component is a type of dropdown menu that consists of a toggle control to open and close a menu of actions or links. Selects differ from dropdowns in that they persist selection where dropdowns are typically used to present a list of actions or links.
+
+<img src="./img/basic-select.png" alt="basic select menu" width="303"/>
+
+1. **Toggle:** The toggle is used to open and close the menu. It takes on all of the behaviors of a button. The toggle label always reflects the currently selected item or may hold a placeholder if nothing is selected.
+
+2. **Menu:** The menu contains a list of options or values. Menus should optimally contain seven (7) or less items for easy scanning by the user. For longer menus, items may be grouped. See the [Menu component guidelines](/components/menu/design-guidelines) for more information about supported options, including use of icons, grouping, and descriptions. The Dropdown component includes many common built-in menu variants. It’s also possible to attach a custom menu by using the [Basic panel variant](/components/select/react#panel).
+
+### Select variations
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/select/select.md:25]: From here to the usage header repeats a lot of what's already been added to the guidelines. I would definitely just merge them somehow

File: packages/v4/src/content/design-guidelines/components/select/select.md [hunk 11]
```diff
@@ -9,7 +9,29 @@ related: [
...
+## Elements and variations
+
+### Basic select
+A select component is a type of dropdown menu that consists of a toggle control to open and close a menu of actions or links. Selects differ from dropdowns in that they persist selection where dropdowns are typically used to present a list of actions or links.
+
+<img src="./img/basic-select.png" alt="basic select menu" width="303"/>
+
+1. **Toggle:** The toggle is used to open and close the menu. It takes on all of the behaviors of a button. The toggle label always reflects the currently selected item or may hold a placeholder if nothing is selected.
+
+2. **Menu:** The menu contains a list of options or values. Menus should optimally contain seven (7) or less items for easy scanning by the user. For longer menus, items may be grouped. See the [Menu component guidelines](/components/menu/design-guidelines) for more information about supported options, including use of icons, grouping, and descriptions. The Dropdown component includes many common built-in menu variants. It’s also possible to attach a custom menu by using the [Basic panel variant](/components/select/react#panel).
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/select/select.md:23]: This is confusing - did you mean to change the second half of the text but forgot to? It seems like it's talking about drop downs even though this is the select guidelines?
I also would want to add a section on scrolling here, or some other recommendations for when the select list is too long. In most case, if someones list is too long, grouping it wouldn't help with that issue! I'd expect a recommendation of making the menu scrollable, or adding a "see more" button at the bottom to view more options. I know this is a common battle/question in c.rh.c

File: packages/v4/src/content/design-guidelines/components/select/select.md (hunk 10 above)
Comment (mmenestr) [reply packages/v4/src/content/design-guidelines/components/select/select.md:25]: Or maybe delete this who select variations - it's a little confusing too because typeahead and checkbox do very different things ! If anything i'd expected a checkbox vs multiple select comparison?

File: packages/v4/src/content/design-guidelines/components/menu/menu.md (hunk 2 above)
Comment (mcarrano) [reply packages/v4/src/content/design-guidelines/components/menu/menu.md:6]: OK. I added an additional sentence. See what you think.

File: packages/v4/src/content/design-guidelines/components/menu/menu.md (hunk 4 above)
Comment (mcarrano) [reply packages/v4/src/content/design-guidelines/components/menu/menu.md:25]: Yeah, in the component examples they are blue. I can fix these, but it should also be fixed in the Sketch library.

File: packages/v4/src/content/design-guidelines/components/menu/menu.md (hunk 8 above)
Comment (mcarrano) [reply packages/v4/src/content/design-guidelines/components/menu/menu.md:71]: This one I'm not sure about. I was using the word "dropdown" to refer to the type of menu generically rather than the component named "Dropdown." Options select menus are used in the Options menu component as well as the Select component. Ultimately we are looking to use the Menu component as part of a composable dropdown menu approach where you could put together custom dropdown menus. What if I just get rid of the work "dropdown" here and just say, "Select from a list of options?"

File: packages/v4/src/content/design-guidelines/components/select/select.md (hunk 11 above)
Comment (mcarrano) [reply packages/v4/src/content/design-guidelines/components/select/select.md:23]: Yes, this is confusing. I did a quick copy and paste without reading this thoroughly. Hopefully it's better now.

File: packages/v4/src/content/design-guidelines/components/select/select.md (hunk 10 above)
Comment (mcarrano) [reply packages/v4/src/content/design-guidelines/components/select/select.md:25]: I'm going to just delete this section. I added it at the last minute, but I see where it is redundant. I mainly wanted to just point out the basic elements and reference the menu guidelines as I do in the first section.

**Comment by mcarrano** [comment]
@mmenestr This is building now. I addressed your comments. Please re-review.

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+3. **Descriptions (optional):** Descriptive text may be added below each menu item. This should only be used when the menu item label by itself may not be clear to all users. Keep descriptive text to two lines or less.
+
+### Grouped menus
+You may decide to group menu items to associate related items and/or to indicate a hierarchy of items. Items may be grouped using group headings, separators, or both.
+
+<img src="./img/grouped-menu.png" alt="grouped menu" width="348"/>
+
+1. **Group heading (optional):** Add a group heading when you want to name the group.
+
+2. **Separator:** Separators are horizontal dividers that help to group the menu items by clearly showing where one group ends and the next begins.
```
[the remark this answers is not in the document]
Comment (mmenestr) [reply packages/v4/src/content/design-guidelines/components/menu/menu.md:40]: @mcarrano Do you disagree with this^?

File: packages/v4/src/content/design-guidelines/components/menu/menu.md (hunk 8 above)
Comment (mmenestr) [reply packages/v4/src/content/design-guidelines/components/menu/menu.md]: Yes, I think that would be a good approach. I just want to steer clear of using the word "dropdown" unless it's referring to the actual dropdown component

File: packages/v4/src/content/design-guidelines/components/select/select.md [hunk 13]
```diff
@@ -9,7 +9,17 @@ related: [
...
 ---
 
-**Select lists** enable users to select one or more items from a list. Select lists should also be used when options are dynamic or variable. 
+**Select lists** enable users to select one or more items from a list. Select lists should also be used when options are dynamic or variable.
+
+## Elements
+
+A select component consists of a toggle control to open and close a menu of actions or links. Selects differ from [dropdowns](/components/dropdown) in that they persist selection where dropdowns are typically used to present a list of actions or links.
+
+<img src="./img/basic-select.png" alt="basic select menu" width="303"/>
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/select/select.md:18]: Need to change this image too to use a blue checkmark instead of grey!

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+3. **Links:** Links may be used in place of or mixed with actions to navigate the user to a new page. If the link will open in a new window, the external link icon should be used to annotate the link and let the user know to expect this behavior.
+
+4. **Descriptions (optional):** Descriptive text may be added below each menu item. This should only be used when the menu item label by itself may not be clear to all users. Keep descriptive text to two lines or less.
+
+### Option select menus
+Use an option select menu when you want to persist the selected items. This is typically the case for select menus or settings menus that allow users to select between multiple options. See the [Select](/components/select) and [Options menu](/components/options-menu) components for more details. You may present a single or multiple [groups of options](#grouped-menus) within the same menu.
+
+<img src="./img/option-select-menu.png" alt="option select menus" width="863"/>
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/menu/menu.md:25]: I think the checkmarks in this image are the wrong blue - Marie updated the symbol now so you should be able to use that now!

**Comment by mmenestr** [comment]
Revisions looked good! Just left a few more comments for things I noticed

File: packages/v4/src/content/design-guidelines/components/menu/menu.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (mcarrano) [reply packages/v4/src/content/design-guidelines/components/menu/menu.md:40]: I can change this to optional @mmenestr . I guess it raises a question of when to use a separator and when not to. I looked at the examples for Select. I think the grouped checkbox select examples are fine without separators, but for other examples it would be hard to see the grouping without them. I think this is partly because the style of the section headers is not that distinct from other menu items. Thoughts?

[the remark this answers is not in the document]
Comment (mmenestr) [reply packages/v4/src/content/design-guidelines/components/menu/menu.md:40]: That's a good point, maybe we say that in dropdown menus it makes sense to use a separator where as in select lists it's not necessary because there is more visual distinction!

File: packages/v4/src/content/design-guidelines/components/menu/menu.md (hunk 14 above)
Comment (mcarrano) [reply packages/v4/src/content/design-guidelines/components/menu/menu.md:25]: Wow. You've got a good eye! Fixed now.

**Comment by mcarrano** [comment]
@mmenestr I made a few more updates. Can you review one more time? Thanks.

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+1. **Selected item:** A checkmark to the right of a menu item indicates that it is currently selected.
+
+2. **Icon (optional):** You may place an icon before each menu item to aid in recognition. It has been shown that familiar icons speed the recognition of text labels, but only use icons if they will be easily recognized and are distinguishable. Never use icons simply as a decoration.
+
+3. **Descriptions (optional):** Descriptive text may be added below each menu item. This should only be used when the menu item label by itself may not be clear to all users. Keep descriptive text to two lines or less.
+
+### Grouped menus
+You may decide to group menu items to associate related items and/or to indicate a hierarchy of items. Items may be grouped using group headings, separators, or both.
+
+<img src="./img/grouped-menu.png" alt="grouped menu" width="348"/>
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/menu/menu.md:36]: This image looks tiny - I think you need to update the width to match the new artboard!

File: packages/v4/src/content/design-guidelines/components/menu/menu.md [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+1. **Group heading (optional):** Add a group heading when you want to name the group.
+
+2. **Separator (optional):** Separators are horizontal dividers that help to group the menu items by clearly showing where one group ends and the next begins. You should consider whether a separator is needed to create visually distinct groups. In the first example above, a separator is used to create separation between the two groups of items. In the second example, The separator is not needed because the titled checkbox groups create two clearly grouped set of options on their own.
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/menu/menu.md:40]: T is un=necessarily capitalize in the last sentence, where it says "The seperator" after the comma

File: packages/v4/src/content/design-guidelines/components/select/select.md [hunk 17]
```diff
@@ -19,22 +29,22 @@ PatternFly offers several variations of select lists:
 * [Multiple select](#multiple-select)
 * [Typeahead](#typeahead)
 
-#### Single select 
+#### Single select
```
Comment (mmenestr) [inline packages/v4/src/content/design-guidelines/components/select/select.md:32]: Realizing now that these headings should probably be made H3s instead of H4s, because right now the table of contents looks very odd

Comment (mcarrano) [reply packages/v4/src/content/design-guidelines/components/select/select.md:32]: Can we open a new issue for that and get this one merged?

**Comment by mcarrano** [comment]
@mmenestr Here it is one more time. Made fixes to the Menu guidelines to address the two issues above.

**Comment by mmenestr** [comment]
@mcarrano the first image in select list still has the grey checkmark - after that should be ready to go! Sorry I didn't notice it yesterday afternoon

**Comment by mcarrano** [comment]
@mmenestr I updated that last image on the Select page and also fixed the heading styles as you suggested while I was there. Please review one more time. Thanks.

**Review by mmenestr — APPROVED** [review]
